org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/wizard/template/ProjectTemplateProcessor.xtend

```diff
@@ -0,0 +1,110 @@
+/*******************************************************************************
+ * Copyright (c) 2017 itemis AG (http://www.itemis.de) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ui.wizard.template
+
+import org.eclipse.xtend.lib.macro.AbstractClassProcessor
+import org.eclipse.xtend.lib.macro.CodeGenerationContext
+import org.eclipse.xtend.lib.macro.TransformationContext
+import org.eclipse.xtend.lib.macro.declaration.ClassDeclaration
+import org.eclipse.xtend.lib.macro.declaration.MutableClassDeclaration
+
+/**
+ * Generate some code to simplify implementation of project templates.
+ * 
+ * <ol>
+ * <li>Automatically extend AbstractProjectTemplate</li>
+ * <li>Generate "messages.properties" for i18n</li>
+ * <li>Generate "Messages.java" for i18n</li>
+ * </ol>
+ * 
+ * The generated files for i18n contain the "label" and "description" of all the project templates. The files may be
+ * extended manually by the user to externalize more strings. The generator then merges its own changes into the
+ * existing files.
+ * 
+ * @author Arne Deutsch - Initial contribution and API
+ * @since 2.14
+ */
+class ProjectTemplateProcessor extends AbstractClassProcessor {
+
+	override doTransform(MutableClassDeclaration annotatedClass, extension TransformationContext context) {
+		annotatedClass.extendedClass = AbstractProjectTemplate.newTypeReference // extend AbstractProjectTemplate
+	}
+
+	override doGenerateCode(ClassDeclaration annotatedClass, extension CodeGenerationContext context) {
+		val propertyContents = generatePropertiesFile(annotatedClass, context) // "messages.properties" for i18n
+		generateMessagesClass(propertyContents, annotatedClass, context) // "Messages.java" for i18n
+	}
+
+	// read the "label" and "description" from the ProjectTemplateAnnotation and update "messages.properties" with the values
+	private def generatePropertiesFile(ClassDeclaration annotatedClass, extension CodeGenerationContext context) {
+		val annotation = annotatedClass.findAnnotation(context.findTypeGlobally(ProjectTemplate))
+		val label = replaceNewlines(annotation.getStringValue("label"))
```

Comment:
This looks like a good use-case for `java.util.Properties.store`, possibly in a variant like `org.eclipse.core.internal.preferences.SortedProperties`